### PR TITLE
hal: telink: tl321x: add Tl321x for TLX Platform based on BLE release v4.0.4.2

### PR DESCRIPTION
**Pull request by damien0x0023**
- update the liblt library file
- update drivers/TL321X, ble/stack/ble/TL321X, ble/vendor/controller/tlx.
- update latest ble lib url in module.yml.
- avoid compile mismatch and issues for PM driver.
- apply compile option -flto=auto to match tl321x ble library.
- adjust TLX_TX_POWER_MIN/TLX_TX_POWER_MAX for tl321x in b9x_rf_power.h.
- apply reg_usb_edp_en |= ep in usbhw.h.
- clean unnecessary comments and codes.
- update crypto/embedtls.

### CODE REVIEW

File: tlsr9/CMakeLists.txt [hunk 1]
```diff
@@ -30,13 +32,20 @@ zephyr_include_directories(drivers/${SOC})
 
 # add for B92 headers
 zephyr_include_directories(drivers/${SOC}/ext_driver)
+if (NOT CONFIG_SOC_RISCV_TELINK_TL321X)
 zephyr_include_directories(drivers/${SOC}/ext_driver/driver_lib)
+else()
+zephyr_include_directories(drivers/${SOC}/ext_driver/driver_internal)
+endif()
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: Better for reading should be next:
```
if (CONFIG_SOC_RISCV_TELINK_TL321X)
zephyr_include_directories(drivers/${SOC}/ext_driver/driver_internal)
else()
zephyr_include_directories(drivers/${SOC}/ext_driver/driver_lib)
endif()
```

File: tlsr9/CMakeLists.txt [hunk 2]
```diff
@@ -30,13 +32,20 @@ zephyr_include_directories(drivers/${SOC})
...
 zephyr_include_directories(drivers/${SOC}/ext_peripherals/codec_0581)
 zephyr_include_directories(drivers/${SOC}/flash)
 
 # soc.c reference sources
+if (NOT CONFIG_SOC_RISCV_TELINK_TL321X)
 zephyr_library_sources(drivers/${SOC}/clock.c)
 zephyr_library_sources(drivers/${SOC}/analog.c)
+zephyr_library_sources(drivers/${SOC}/stimer.c)
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: Why we added here "stimer.c" ?
This change related for all SoCs excluding TLX, if it works before, maybe "stimer.c" not needed to add here

File: tlsr9/CMakeLists.txt [hunk 3]
```diff
@@ -70,50 +89,88 @@ if(CONFIG_BT_B9X OR CONFIG_IEEE802154_TELINK_B9X)
...
+	ble/common/utility.c
+	ble/vendor/controller/tlx/tlx_bt.c
+	ble/vendor/controller/tlx/tlx_bt_init.c
+	ble/vendor/controller/tlx/tlx_bt_buffer.c
+	ble/vendor/controller/tlx/tlx_bt_flash.c
+)
+
+endif() # (CONFIG_BT_B9X OR CONFIG_BT_TLX)
+
+add_definitions(-DSTD_GCC)
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: is this change necessary ?

File: tlsr9/common/b9x_rf_power.c
Comment (s07641069) [inline tlsr9/common/b9x_rf_power.c:1]: This file now is common for B9X and TLX
Better to rename, remove "b9x_" from file name
during compilation you will see where file used to fix compilation.

File: tlsr9/common/b9x_rf_power.h
Comment (s07641069) [inline tlsr9/common/b9x_rf_power.h:1]: This file now is common for B9X and TLX
Better to rename, remove "b9x_" from file name
during compilation you will see where file used to fix compilation.

File: tlsr9/common/b9x_sleep.h
Comment (s07641069) [inline tlsr9/common/b9x_sleep.h:1]: This file now is common for B9X and TLX
Better to rename, remove "b9x_" from file name
during compilation you will see where file used to fix compilation.

File: tlsr9/CMakeLists.txt [hunk 4]
```diff
@@ -70,50 +89,88 @@ if(CONFIG_BT_B9X OR CONFIG_IEEE802154_TELINK_B9X)
...
 )
 
-endif() # CONFIG_BT_B9X
+zephyr_library_sources_ifdef(CONFIG_BT_TLX
+	ble/common/utility.c
+	ble/vendor/controller/tlx/tlx_bt.c
+	ble/vendor/controller/tlx/tlx_bt_init.c
+	ble/vendor/controller/tlx/tlx_bt_buffer.c
+	ble/vendor/controller/tlx/tlx_bt_flash.c
+)
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: Here make sense to rename ble controller folders same as for drivers to use ${SOC} and simplify code
For files "b9x_bt*, tlx_bt*" you can use ${SOC} as well, but need to rename files or just remove b9x_ and tlx_ from files name to have more simple code as well. Its up to you.

```
if (CONFIG_BT_B9X OR CONFIG_BT_TLX)

zephyr_include_directories(
	ble
	ble/vendor/controller/${SOC}
)

zephyr_library_sources(
	ble/common/utility.c
	ble/vendor/controller/${SOC}/bt.c
	ble/vendor/controller/${SOC}/bt_init.c
	ble/vendor/controller/${SOC}/bt_buffer.c
	ble/vendor/controller/${SOC}/bt_flash.c
)

endif() # (CONFIG_BT_B9X OR CONFIG_BT_TLX)
```

File: tlsr9/CMakeLists.txt [hunk 5]
```diff
@@ -70,50 +89,88 @@ if(CONFIG_BT_B9X OR CONFIG_IEEE802154_TELINK_B9X)
...
-zephyr_include_directories(
+zephyr_include_directories_ifdef(CONFIG_IEEE802154_TELINK_B9X
 	ble
-	ble/vendor/controller
+	ble/vendor/controller/b9x
+)
+
+zephyr_include_directories_ifdef(CONFIG_IEEE802154_TELINK_TLX
+	ble
+	ble/vendor/controller/tlx
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: Same here, we can use simplified code after some folders/files renaming
```
zephyr_include_directories(
	ble
	ble/vendor/controller/${SOC}
)
```

File: tlsr9/CMakeLists.txt [hunk 6]
```diff
@@ -50,17 +59,27 @@ if(CONFIG_SOC_FLASH_TELINK_B9X OR CONFIG_USB_TELINK_B9X)
...
 zephyr_library_sources_ifdef(CONFIG_PWM_TELINK_B9X drivers/${SOC}/pwm.c)
+zephyr_library_sources_ifdef(CONFIG_PWM_TELINK_TLX drivers/${SOC}/pwm.c)
 
 # SPI driver reference sources
 zephyr_library_sources_ifdef(CONFIG_SPI_TELINK_B9X drivers/${SOC}/gpio.c)
+zephyr_library_sources_ifdef(CONFIG_SPI_TELINK_TLX drivers/${SOC}/gpio.c)
 
 # I2C driver reference sources
 zephyr_library_sources_ifdef(CONFIG_I2C_TELINK_B9X drivers/${SOC}/i2c.c)
+zephyr_library_sources_ifdef(CONFIG_I2C_TELINK_TLX drivers/${SOC}/i2c.c)
```
Comment (s07641069) [inline tlsr9/CMakeLists.txt]: Same here and in all file
investigate if better to change or leave as it is, from one side in some place we save line of code and make code simple and in others it will change two line into tree
Will be good to have one common solution for all includes in this Make file
for example 
if(CONFIG_GPIO_TELINK_B9X OR CONFIG_GPIO_TELINK_TLX)
	zephyr_library_sources(drivers/${SOC}/gpio.c)
endif()

**Review by s07641069 — CHANGES_REQUESTED** [review]
As well need to squash all commits into one 
But this you can do after all review changes will be fixed

File: tlsr9/CMakeLists.txt (hunk 3 above)
Comment (liucoldstar) [reply tlsr9/CMakeLists.txt]: Yes. This macro is provided by the Driver SDK team. It must be provided when using non-Andes GCC.

File: tlsr9/CMakeLists.txt (hunk 1 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: Yes, will change it for better readability.

File: tlsr9/CMakeLists.txt (hunk 2 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: I check with this here (3a3bc5b991d0d8b25b9199eaae212851e5cc3dcc).
We might add the stimer.c for the old file structure of the last TL321x driver version. 
We should remove it because this file has been moved into our BLE lib in the new file structure.

File: tlsr9/CMakeLists.txt (hunk 3 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: Since `-DSTD_GCC` is only valid for TL321X, we may add conditional lines like this

 `if (CONFIG_SOC_RISCV_TELINK_TL321X)`
  `add_definitions(-DSTD_GCC)`
`endif()`

File: tlsr9/common/b9x_rf_power.c
Comment (damien0x0023) [reply tlsr9/common/b9x_rf_power.c:1]: @s07641069 
Would it be better to create new files for the TLX family only?
For example,
`tlsr9/common/tlx_rf_power.c`
`tlsr9/common/tlx_rf_power.h`
`tlsr9/common/tlx_sleep.c`
`tlsr9/common/tlx_sleep.h`

Or we could rename their pre-fix from  `b9x` to `tl_` for `telink`.
For example,
`tlsr9/common/tl_context.c`
`tlsr9/common/tl_rf_power.c`
`tlsr9/common/tl_rf_power.h`
`tlsr9/common/tl_sleep.c`
`tlsr9/common/tl_sleep.h`

File: tlsr9/CMakeLists.txt (hunk 4 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: ${SOC} are `B91`, `B92`, `B95`, `TL321X`.
However, there are only two sub-folders `b9x` and `tlx` under the `controller` folder.
I wonder if I can add
`set(SOC "TL321X")` at the beginning of the file,
so I can include files from 
`ble/vendor/controller/b9x/b9x_bt.c` or `ble/vendor/controller/tlx/tlx_bt.c`
to
`ble/vendor/controller/${SOC_SERIES}/${SOC_SERIES}_bt.c`

File: tlsr9/common/b9x_rf_power.c
Comment (s07641069) [reply tlsr9/common/b9x_rf_power.c:1]: @damien0x0023 In case if for TLX we don't need separate files and we can use same as for B9X better to avoid files duplication

File: tlsr9/CMakeLists.txt (hunk 4 above)
Comment (s07641069) [reply tlsr9/CMakeLists.txt]: ok this is up to you with some optimizations

File: tlsr9/common/b9x_rf_power.c
Comment (damien0x0023) [reply tlsr9/common/b9x_rf_power.c:1]: @andy-liu-telink @s07641069 
All variables and APIs in `tlsr9/ble/vendor/controller/tlx/tlx_bt.c` and so on are still with `b9x_`.
Thus, I created separate files for the TLX SoC family and modified that' tlx _` to `b9x_` instead.

File: tlsr9/ble/vendor/controller/Kconfig [hunk 7]
```diff
@@ -1,9 +1,8 @@
 # Copyright (c) 2022 Telink Semiconductor
 # SPDX-License-Identifier: Apache-2.0
 
-if BT_B9X
-
 menu "B9X BLE Controller"
+	depends on BT_B9X # Only show B9X menu if BT_B9X is enabled
```
Comment (s07641069) [inline tlsr9/ble/vendor/controller/Kconfig]: do not remove if BT_B9X
and add if BT_TLX 
otherwise all this B9X configs and TLX will be available for both SoC family in .config during build

File: tlsr9/common/b9x_sleep.h [hunk 8]
```diff
@@ -33,6 +33,6 @@
 bool b9x_suspend(uint32_t wake_stimer_tick);
 #if CONFIG_SOC_SERIES_RISCV_TELINK_B9X_RETENTION
 bool b9x_deep_sleep(uint32_t wake_stimer_tick);
-#endif /* CONFIG_SOC_SERIES_RISCV_TELINK_B9X_RETENTION */
+#endif /* (CONFIG_SOC_SERIES_RISCV_TELINK_B9X_RETENTION */
```
Comment (s07641069) [inline tlsr9/common/b9x_sleep.h]: revert 
use original 
`#endif /* CONFIG_SOC_SERIES_RISCV_TELINK_B9X_RETENTION */`

**Review by s07641069 — CHANGES_REQUESTED** [review]

**Review by s07641069 — APPROVED** [review]

**Review by s07641069 — CHANGES_REQUESTED** [review]
do not forget to squash all commits after all fixes

File: tlsr9/common/b9x_rf_power.c
Comment (damien0x0023) [reply tlsr9/common/b9x_rf_power.c:1]: Rename all `b9x_` and `tlx_` to `tl_` here and under the directory `tlsr9/ble/vendor/controller/tlx/`.

Comment (damien0x0023) [reply tlsr9/common/b9x_rf_power.c:1]: `b9x_rf_power.c` has been rename to `tl_rf_power.c` for both SoC families.

File: tlsr9/common/b9x_rf_power.h
Comment (damien0x0023) [reply tlsr9/common/b9x_rf_power.h:1]: Has renamed `b9x_rf_power.h` to `tl_rf_power.h`.

File: tlsr9/common/b9x_sleep.h
Comment (damien0x0023) [reply tlsr9/common/b9x_sleep.h:1]: Has renamed `b9x_sleep.h` to `tl_sleep.h`.

File: tlsr9/CMakeLists.txt (hunk 4 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: updated using `string(REGEX REPLACE "^telink_" "" SOC_FAMILY ${CONFIG_SOC_SERIES})`

File: tlsr9/ble/vendor/controller/Kconfig (hunk 7 above)
Comment (damien0x0023) [reply tlsr9/ble/vendor/controller/Kconfig]: Rename choices and configs from `B9X_` to `TL_` for both SoC families.

File: tlsr9/CMakeLists.txt (hunk 5 above)
Comment (damien0x0023) [reply tlsr9/CMakeLists.txt]: use ${SOC_FAMILY} for `b9x` and `tlx` instead.

File: tlsr9/common/b9x_sleep.h (hunk 8 above)
Comment (damien0x0023) [reply tlsr9/common/b9x_sleep.h]: renamed b9x_sleep.h to tl_sleep.h for both `b9x` and `tlx` in latest commit 0941ddd.

**Comment by damien0x0023** [comment]
> do not forget to squash all commits after all fixes

Yes. Rebase the branch and squash all commits into one.

**Review by s07641069 — APPROVED** [review]
